crates/ide_assists/src/handlers/generate_documentation_template.rs

```diff
@@ -0,0 +1,1169 @@
+use ide_db::assists::{AssistId, AssistKind};
+use stdx::to_lower_snake_case;
+use syntax::{
+    ast::{self, edit::IndentLevel, HasDocComments, HasName},
+    AstNode,
+};
+
+use crate::assist_context::{AssistContext, Assists};
+
+// Assist: generate_documentation_template
+//
+// Adds a documentation template above a function definition / declaration.
+//
+// ```
+// pub fn my_$0func(a: i32, b: i32) -> Result<(), std::io::Error> {
+//     unimplemented!()
+// }
+// ```
+// ->
+// ```
+// /// .
+// ///
+// /// # Examples
+// ///
+// /// ```
+// /// use test::my_func;
+// ///
+// /// assert_eq!(my_func(a, b), );
+// /// ```
+// ///
+// /// # Errors
+// ///
+// /// This function will return an error if .
+// pub fn my_func(a: i32, b: i32) -> Result<(), std::io::Error> {
+//     unimplemented!()
+// }
+// ```
+pub(crate) fn generate_documentation_template(
+    acc: &mut Assists,
+    ctx: &AssistContext,
+) -> Option<()> {
+    let name = ctx.find_node_at_offset::<ast::Name>()?;
+    let ast_func = name.syntax().parent().and_then(ast::Fn::cast)?;
+    if is_in_trait_impl(&ast_func, ctx)
+        || !is_public(&ast_func, ctx)?
+        || ast_func.doc_comments().next().is_some()
+    {
+        return None;
+    }
+
+    let parent_syntax = ast_func.syntax();
+    let text_range = parent_syntax.text_range();
+    let indent_level = IndentLevel::from_node(&parent_syntax);
+
+    acc.add(
+        AssistId("generate_documentation_template", AssistKind::Generate),
+        "Generate a documentation template",
+        text_range,
+        |builder| {
+            let mut doc_lines = Vec::new();
+            // Introduction / short function description before the sections
+            doc_lines.push(introduction_builder(&ast_func));
+            // Then come the sections
+            if let Some(mut lines) = examples_builder(&ast_func, ctx) {
+                doc_lines.push("".into());
+                doc_lines.append(&mut lines);
+            }
+            for section_builder in [panics_builder, errors_builder, safety_builder] {
+                if let Some(mut lines) = section_builder(&ast_func) {
+                    doc_lines.push("".into());
+                    doc_lines.append(&mut lines);
+                }
+            }
+            builder.insert(text_range.start(), documentation_from_lines(doc_lines, indent_level));
+        },
+    )
+}
+
+/// Builds an introduction, trying to be smart if the function is `::new()`
+fn introduction_builder(ast_func: &ast::Fn) -> String {
+    let is_new = ast_func.name().map(|name| &name.to_string() == "new").unwrap_or(false);
+    if is_new {
+        let ret_type = return_type(ast_func).map(|ret_type| ret_type.to_string());
+        let self_type = self_type(ast_func);
+        if ret_type.as_deref() == Some("Self") || ret_type == self_type {
+            if let Some(self_type) = self_type {
+                return format!("Creates a new [`{}`].", self_type);
+            }
+        }
```

Comment:
This part should probably go through `hir` as well, that is fetch the name of things through the hir instead of doing it text based, `fn_def.ret_type(db).as_adt().name(db)`. Similarly could you then check if the function returns the same type as the impl by comparing the hir types instead of comparing text from the AST which could be different but still resolve to the same type.

Reply:
Done, but in the pushed version the `ast` is still used to get the name of the type, because the `hir` does not add the generic types `<T>` into the name and they should appear in the documentation text. Is there a way to get this with `hir`?

Reply:
You can get a rendered string of a hir type by using the `HirDisplay` trait, https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/hir_ty/src/display.rs#L48

Reply:
It's okay to mix hir and ast for assists as some things are easier in one of the two than the other. But usually if possible using hir is preferred as its more structured.

It's fine to keep this on the AST if it is too troublesome.

Reply:
IIUC it requires to define a module to display code that can be compiled, so it seems quite troublesome :sweat_smile:

Reply:
No worries!